src/rust/inetstack/protocols/tcp/peer.rs

```diff
@@ -295,196 +339,211 @@ impl TcpPeer {
-            None => return Err(Fail::new(EBADF, "bad queue descriptor")),
-        }
-
-        Ok(())
+            _ => return Err(Fail::new(libc::EBADF, "bad queue descriptor")),
+        };
+        inner.addresses.remove(&SocketId::Passive(addr));
+        Err(Fail::new(libc::ENOTSUP, "close not implemented for listening sockets"))
     }
 
-    pub fn remote_mss(&self, fd: QDesc) -> Result<usize, Fail> {
+    pub fn remote_mss(&self, qd: QDesc) -> Result<usize, Fail> {
+        #[cfg(feature = "profiler")]
+        timer!("tcp::remote_mss");
         let inner = self.inner.borrow();
-        let key = match inner.sockets.get(&fd) {
-            Some(Socket::Established { local, remote }) => (*local, *remote),
-            Some(..) => return Err(Fail::new(ENOTCONN, "connection not established")),
-            None => return Err(Fail::new(EBADF, "bad queue descriptor")),
-        };
-        match inner.established.get(&key) {
-            Some(ref s) => Ok(s.remote_mss()),
-            None => Err(Fail::new(ENOTCONN, "connection not established")),
+        let qtable: Ref<IoQueueTable<InetQueue>> = inner.qtable.borrow();
+        match qtable.get(&qd) {
+            Some(InetQueue::Tcp(queue)) => match queue.get_socket() {
+                Socket::Established(socket) => Ok(socket.remote_mss()),
+                _ => Err(Fail::new(libc::ENOTCONN, "connection not established")),
+            },
+            _ => Err(Fail::new(libc::EBADF, "bad queue descriptor")),
         }
     }
 
-    pub fn current_rto(&self, fd: QDesc) -> Result<Duration, Fail> {
+    pub fn current_rto(&self, qd: QDesc) -> Result<Duration, Fail> {
+        #[cfg(feature = "profiler")]
+        timer!("tcp::current_rto");
         let inner = self.inner.borrow();
-        let key = match inner.sockets.get(&fd) {
-            Some(Socket::Established { local, remote }) => (*local, *remote),
-            Some(..) => return Err(Fail::new(ENOTCONN, "connection not established")),
-            None => return Err(Fail::new(EBADF, "bad queue descriptor")),
-        };
-        match inner.established.get(&key) {
-            Some(ref s) => Ok(s.current_rto()),
-            None => Err(Fail::new(ENOTCONN, "connection not established")),
+        let qtable: Ref<IoQueueTable<InetQueue>> = inner.qtable.borrow();
+        match qtable.get(&qd) {
+            Some(InetQueue::Tcp(queue)) => match queue.get_socket() {
+                Socket::Established(socket) => Ok(socket.current_rto()),
+                _ => return Err(Fail::new(libc::ENOTCONN, "connection not established")),
+            },
+            _ => return Err(Fail::new(libc::EBADF, "bad queue descriptor")),
         }
     }
 
-    pub fn endpoints(&self, fd: QDesc) -> Result<(SocketAddrV4, SocketAddrV4), Fail> {
+    pub fn endpoints(&self, qd: QDesc) -> Result<(SocketAddrV4, SocketAddrV4), Fail> {
+        #[cfg(feature = "profiler")]
+        timer!("tcp::endpoints");
```

Comment:
Could we add changes that concern performance observation in a different commit?

Reply:
I thought I would go ahead and do this since I was also fixing the fd -> qd issues.